legacy_status_collector.go

```diff
@@ -0,0 +1,174 @@
+package statusresource
+
+import (
+	"sync"
+	"time"
+
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var (
+	legacyStatusCollectorDescription *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName("statusresource", "cluster", "status"),
+		"Cluster status condition as provided by the CR status.",
+		[]string{
+			"cluster_id",
+			"status",
+		},
+		nil,
+	)
+	legacyCreationSecondsCollectorDescription *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName("statusresource", "cluster", "creation_seconds"),
+		"Cluster creation duration.",
+		[]string{
+			"cluster_id",
+		},
+		nil,
+	)
+)
+
+type LegacyStatusCollectorConfig struct {
+	Logger  micrologger.Logger
+	Watcher func(opts metav1.ListOptions) (watch.Interface, error)
+}
+
+type LegacyStatusCollector struct {
+	logger  micrologger.Logger
+	watcher func(opts metav1.ListOptions) (watch.Interface, error)
+
+	bootOnce sync.Once
+}
+
+func NewLegacyStatusCollector(config LegacyStatusCollectorConfig) (*LegacyStatusCollector, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Watcher == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Watcher must not be empty", config)
+	}
+
+	c := &LegacyStatusCollector{
+		logger:  config.Logger,
+		watcher: config.Watcher,
+
+		bootOnce: sync.Once{},
+	}
+
+	return c, nil
+}
+
+func (c *LegacyStatusCollector) Collect(ch chan<- prometheus.Metric) error {
+	watcher, err := c.watcher(metav1.ListOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	defer watcher.Stop()
+
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				continue
+			}
+
+			m, err := meta.Accessor(event.Object)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			p, ok := event.Object.(Provider)
+			if !ok {
+				panic("asserting Provider interface failed")
+			}
+
+			if p.ClusterStatus().HasCreatedCondition() {
+				t1 := timeForCreatedCondition(p)
+				t2 := m.GetCreationTimestamp().Time
+
+				ch <- prometheus.MustNewConstMetric(
+					legacyCreationSecondsCollectorDescription,
+					prometheus.GaugeValue,
+					t1.Sub(t2).Seconds(),
```

Comment:
Could it be summary or histogram to see how it distributed?

Reply:
I think a histogram only makes sense for changing values. The creation time does never change once emitted.

Reply:
Thanks, that make sense.